### PR TITLE
APPS-1808 Nextflow | Reduce API requests related to checking subjob error strategy

### PR DESCRIPTION
**Pull request by r-i-v-a**

### CODE REVIEW

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 1]
```diff
@@ -442,53 +448,58 @@ main() {
...
-  else
-    >&2 echo "Missing Nextflow .command.log file"
-  fi
-  # mark the job as successful in any case, real task
-  # error code is managed by nextflow via .exitcode file
-  if [ -z ${exit_code} ]; then export exit_code=0; fi
-
-  # Make sure that subjob with errorStrategy == terminate end in 'failed' state
-  wait_time=240
+wait_for_terminate_or_retry() {
```
Comment (r-i-v-a) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:451]: Refactor -- separate wait_for_terminate_or_retry from nf_task_exit

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 2]
```diff
@@ -9,6 +9,12 @@ DOCKER_CREDS_FOLDER=/docker/credentials/
 DOCKER_CREDS_FILENAME=dx_docker_creds
 CREDENTIALS=${HOME}/credentials
 
+# How long to let a subjob with error keep running for Nextflow to handle it
+# before we end the DX job, in seconds
+MAX_WAIT_AFTER_JOB_ERROR=240
+# How often to check when waiting for a subjob with error, in seconds
+WAIT_INTERVAL=15
```
Comment (r-i-v-a) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:16]: Increased interval from 10s --> 15s

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh (hunk 1 above)
Comment (r-i-v-a) [reply src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:451]: Only look for the error strategy files in the workspace _if_ in the condition where the subjob had error

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 3]
```diff
@@ -442,53 +448,58 @@ main() {
...
+  retry_record=$(dx find data --name $DX_JOB_ID --path $DX_WORKSPACE_ID:/.RETRY --brief | head -n 1)
+  if [ -n "${retry_record}" ]; then
     wait_period=0
     echo "Subjob exited with non-zero exit_code and the errorStrategy is retry."
-    echo "Waiting for the headjob to kill the job tree or for instruction to continue"
-
+    echo "Waiting for the head job to kill the job tree or for instruction to continue..."
     while true
     do
-        dx describe $DX_JOB_ID --json | jq .properties -r
```
Comment (r-i-v-a) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:476]: This dx describe call wasn't used

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 4]
```diff
@@ -442,53 +448,58 @@ main() {
...
     echo "Subjob exited with non-zero exit_code and the errorStrategy is terminate."
-    echo "Waiting for the headjob to kill the job tree..."
-    sleep $wait_time
+    echo "Waiting for the head job to kill the job tree..."
+    sleep $MAX_WAIT_AFTER_JOB_ERROR
     echo "This subjob was not killed in time, exiting to prevent excessive waiting."
-    exit
+    # Expected the job tree to be killed because of "terminate" strategy
+    # If not done already by Nextflow, will do it here
+    exit -1
```
Comment (r-i-v-a) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:460]: `exit -1` see APPS-1821

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 5]
```diff
@@ -442,53 +448,58 @@ main() {
     exit $ret
 }
 
-# On exit, for the Nextflow task sub-jobs
-nf_task_exit() {
-  ret=$?
```
Comment (r-i-v-a) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:447]: Removed unused variable `ret`

**Comment by r-i-v-a** [comment]
Nextflow QE tests succeeded
https://github.com/dnanexus/nextaur-app/actions/runs/4800431270

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 6]
```diff
@@ -442,53 +448,58 @@ main() {
...
     echo "Subjob exited with non-zero exit_code and the errorStrategy is terminate."
-    echo "Waiting for the headjob to kill the job tree..."
-    sleep $wait_time
+    echo "Waiting for the head job to kill the job tree..."
+    sleep $MAX_WAIT_AFTER_JOB_ERROR
     echo "This subjob was not killed in time, exiting to prevent excessive waiting."
-    exit
+    # Expected the job tree to be killed because of "terminate" strategy
+    # If not done already by Nextflow, will do it here
+    exit $exit_code
```
Comment (r-i-v-a) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:460]: This is my opinion about how to handle this situation; open to discussion. APPS-1821

**Comment by r-i-v-a** [comment]
Open to comments on draft @commandlinegirl @YuxinShi0423 @mhrvol

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh (hunk 6 above)
Comment (mhrvol) [reply src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:460]: I added comment to the jira ticket, we can go through it during our daily if you want

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh [hunk 7]
```diff
@@ -442,53 +448,58 @@ main() {
...
+    dx upload .command.log --path "${cmd_log_file}" --brief --wait --no-progress || true
+  else
+    >&2 echo "Missing Nextflow .command.log file"
+  fi
+
+  # exit_code should already be set in nf_task_entry(); default just in case
+  # This is just for including as DX output; Nextflow internally uses .exitcode file
+  if [ -z ${exit_code} ]; then export exit_code=0; fi
+
+  if [ "$exit_code" -ne 0 ]; then wait_for_terminate_or_retry; fi
```
Comment (mhrvol) [inline src/python/dxpy/templating/templates/nextflow/src/nextflow.sh:498]: Thanks, much clearer after putting it to a separate function.

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,17 @@ Categories for each release: Added, Changed, Deprecated, Removed, Fixed, Securit
...
+### Changed
+
+* Optimizations in Nextflow Pipeline Applet script to make fewer API calls when
+concluding a subjob
+
+### Fixed
+
+* If Nextflow subjob finished with error and error strategy was terminate, wait up to 240s
+then exit the subjob with error, which will cause the pipeline exec tree to fail
+(consistent with Nextflow's expected behavior if not timed out).
```
Comment (mhrvol) [inline CHANGELOG.md:18]: It is not true that this is consistent. It's just inconsistent in a different manner. I've added comment to the jira ticket mentioned below.

**Comment by r-i-v-a** [comment]
Nextflow QE tests passed
https://github.com/dnanexus/nextaur-app/actions/runs/4810379556

File: src/python/dxpy/templating/templates/nextflow/src/nextflow.sh (hunk 6 above)
Comment (r-i-v-a) [reply src/python/dxpy/templating/templates/nextflow/src/nextflow.sh]: After discussion, we decided to remove any unrelated logical change from this PR and handle it as a separate Jira / PR. @mhrvol

**Review by commandlinegirl — APPROVED** [review]
LGTM, thanks!

**Pull request merged**